### DISCUSSION TITLE
(yet unknown) barriers lead to illogical network entry point

### DISCUSSION THREAD

**Discussion by tobwen** [valhalla/valhalla#4989] [Q&A]
### Description
As can be seen in the figure, the entry point into the network is the shortest point, but also the most illogical point. Normally you would exclude the direct route that crosses the water. The reason is obvious: the search for the shortest distance is not “penalized” because there are no barriers dataset for this shortest distance.

![image](https://github.com/user-attachments/assets/746a233f-b03f-4f44-aa2a-bdb7b8856f29)
Data: © OpenStreetMap Contributors, license: ODbL v1.0

### Workaround: pre-processing
As a workaround, I have now loaded all natural barriers into PostGIS and determined the nearest points in the network that are not intersected by a barrier for all my entry coordinates. I think I'll make a service out of this, but the calculation is of course expensive.

### Questions
- How do you do it?
- How WOULD you do it?

**Comment by chrstnbwnkl** [comment]
Valhalla does not have a notion of natural barriers. It sounds like a bit of a niche feature to me, as the number of natural barriers that really stand in the way to get to the nearest road is rather small (the Fulda is a good example though), but I see how it can be useful in some off road situations where you can't assume the user is actually close to a given road.

If I were to implement this in Valhalla, I would probably try  to squeeze this data into the graph tiles, similar to how edges are stored in the graphtile (we have one fixed sized structure called `DirectedEdge` that stores an offset into a list of variable sized data structures callled `EdgeInfo`, which also holds the shape for a given edge). This would only work if the number of objects that count as natural barriers is relatively low, so not every tiny stream should count as one, otherwise this would blow up the graph size. the tile grid already works as some kind of _very_ coarse spatial index, but if needed (i.e. if there is a considerate number of natural barrier per tile), you could easily make this a bit more fine grained (tile size is ~27 km at the equator IIRC, you could divide this up into 4x4 subdivisions, which would add 16 bits per natural barrier and give you a ~7 km grid). 

Not saying this should be a feature of public Valhalla, just as an idea of how to effectively perform these checks when searching for edge candidates.

**Reply by chrstnbwnkl** [reply]
Another possibility would be to accept these features per request, similar to `exclude_polygons`, but we'd accept linear features, optionally even with penalties specified per feature. It would leave the data side to the end user, but in term it would be much more generic and in that sense maybe useful for the public project.

**Comment by chrstnbwnkl** [comment]
Actually, have you tried using `exclude_locations`? Maybe this is actually good enough 😄 Valhalla performs a location search for those, just like it does for `locations`. how you sample those linear features would matter of course, but I could imagine you could get pretty far with that method

**Reply by tobwen** [reply]
I haven't tried it yet, as I assume that this straight line is not part of the route. I'll give it a try in the next few days.
